### ISSUE TITLE
Adding custom headers & unhandled rejection on 403

### ISSUE THREAD

**Issue by fergonco** [marmelab/react-admin#5322]
Hi, I am using the example to add custom headers from the json server provider: https://github.com/marmelab/react-admin/tree/master/packages/ra-data-json-server#adding-custom-headers
I am getting a 403 status code from the server and I would expect the error message being notified and nothing else.
However, in addition to the error notification, an error appears in the console:
```
Unhandled Rejection (HttpError): <the error message from the 403 response>

  3 | import {fetchUtils} from 'react-admin';
  4 | 
  5 | const httpClient = (url, options = {}) => {
> 6 |     return fetchUtils.fetchJson(url, options);
  7 | };
  8 | const dataProvider = jsonServerProvider(host, httpClient)
  9 | 
```
It does not happen if I invoke *jsonServerProvider* like this `const dataProvider = jsonServerProvider(host, fetchUtils.fetchJson)` or like this `const dataProvider = jsonServerProvider(host)`.

I fail to see why the rejection of the promise returned by the fetchUtils.fetchJson is handled and if I pass a function returning that same promise it is not handled. But at the very least there should be a comment about this behavior in the documentation.

BTW, thanks for the great project.

**Comment by fzaninotto** [comment]
You need an `authProvider` to log the user out in case of 403 error. 

This is explained in [the Authentication chapter](https://marmelab.com/react-admin/Authentication.html#catching-authentication-errors-on-the-api) of the documentation.

**Comment by fergonco** [comment]
That seem to conflict with other issues were it is said that we can handle HTTP error codes as we want (https://github.com/marmelab/react-admin/issues/974). If I understood correctly of course.

In our case, we want to keep the user logged in the case of 403 because that means for us that the user just attempted a forbidden operation, but he can still attempt others he is allowed to. Our *checkError* method therefore looks like this:
```
   checkError: ({status}) => {
        if (status === 401) {
                return Promise.reject();
        }
        return Promise.resolve();
    },
```

**Comment by fzaninotto** [comment]
In your opinion, what should users see if they try to access a resource that they are not allowed to see?

**Comment by fergonco** [comment]
In my case the server is returning a json body with a *message* property and I was expecting it to be shown in a temporal notification... and indeed it does! The problem is that, in addition to that, an error takes place as well, and in development mode it shows the typical pink/red screen.

Based on the "Error format" section of the docs (https://marmelab.com/react-admin/DataProviders.html#error-format) I would assume that, as the status code is 403, my *AuthProvider.checkError* would be called and, as it returns a resolved promise, nothing else would happen.

**Comment by fergonco** [comment]
Sorry, I think I have misguided the question. The described behavior happens as well with other error status code (404, 500) so it does not seem to be related to authentication.

I have created a project in GitHub (https://github.com/fergonco/ra-vs-fergonco) to reproduce the behavior. Basically it's the setup of the tutorial + the *fetchUtils* wrapper + an edit component for users. When updating an user and getting the error response (404, 403, 500, etc.), the React dev error page will appear if the *fetchUtils* wrapper is used:
```
const httpClient = (url, options = {}) => {
    return fetchUtils.fetchJson(url, options);
};

const dataProvider = jsonServerProvider('http://localhost:3000', httpClient);
```
The error page will not appear if we pass directly `fetchUtils.fetchJson`.
```
const httpClient = (url, options = {}) => {
    return fetchUtils.fetchJson(url, options);
};

const dataProvider = jsonServerProvider('http://localhost:3000', fetchUtils.fetchJson);
```
